### ISSUE TITLE
spurious coherence error using error-chain with nonexistent type

### ISSUE THREAD

**Issue by durka** [rust-lang/rust#43400]
~~~I haven't been able to minimize this, but the code is:~~~ _(edit: see below)_

```rust
error_chain! {
    foreign_links {
        Bad(nope::Thing); // nonexistent type
    }
}
```

The first error is sensical:

```
error[E0433]: failed to resolve. Use of undeclared type or module `nope`
  --> /Users/alex/.cargo/.cargo/script-cache/expr-698cc781c70734ef/expr.rs:11:47
   |
11 |     match {error_chain! { foreign_links { Bad(nope::Thing); } }} {
   |                                               ^^^^^^^^^^^ Use of undeclared type or module `nope`
```

But the knock-on errors are not:

```
error[E0119]: conflicting implementations of trait `std::convert::From<[type error]>` for type `try_main::Error`:
  --> /Users/alex/.cargo/.cargo/script-cache/expr-698cc781c70734ef/expr.rs:11:12
   |
11 |     match {error_chain! { foreign_links { Bad(nope::Thing); } }} {
   |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |            |
   |            first implementation here
   |            conflicting implementation for `try_main::Error`
   |
   = note: this error originates in a macro outside of the current crate
```

**Comment by durka** [comment]
Minimized:

```rust

struct ErrorKind;
struct Error(ErrorKind);

impl From<nope::Thing> for Error {
    fn from(_: nope::Thing) -> Self {
        unimplemented!()
    }
}

impl From<ErrorKind> for Error {
    fn from(_: ErrorKind) -> Self {
        unimplemented!()
    }
}
```

```
error[E0433]: failed to resolve. Use of undeclared type or module `nope`
 --> expanded.rs:7:11
  |
7 | impl From<nope::Thing> for Error {
  |           ^^^^^^^^^^^ Use of undeclared type or module `nope`

error[E0433]: failed to resolve. Use of undeclared type or module `nope`
 --> expanded.rs:8:16
  |
8 |     fn from(_: nope::Thing) -> Self {
  |                ^^^^^^^^^^^ Use of undeclared type or module `nope`

error[E0119]: conflicting implementations of trait `std::convert::From<[type error]>` for type `Error`:
  --> expanded.rs:13:1
   |
7  | / impl From<nope::Thing> for Error {
8  | |     fn from(_: nope::Thing) -> Self {
9  | |         unimplemented!()
10 | |     }
11 | | }
   | |_- first implementation here
12 |
13 | / impl From<ErrorKind> for Error {
14 | |     fn from(_: ErrorKind) -> Self {
15 | |         unimplemented!()
16 | |     }
17 | | }
   | |_^ conflicting implementation for `Error`
```
